server/src/main/java/org/opensearch/search/profile/fetch/FlatFetchProfileTree.java

```diff
@@ -59,8 +60,8 @@ private static class Node {
     }
 
     private final List<Node> roots = new ArrayList<>();
-    private final Map<String, Node> rootsMap = new HashMap<>();
-    private final Map<String, Node> phaseMap = new HashMap<>();
+    private final ConcurrentMap<String, Node> rootsMap = new ConcurrentHashMap<>();
```

Comment:
The following code (from line 71) is not thread safe:

```java
Node node = rootsMap.get(uniqueElement);
if (node == null) {
    node = new Node(element); // Keep original element name for display
    roots.add(node);
    rootsMap.put(uniqueElement, node);
}
```

Two threads could end up creating a new node but one of them will be overwritten. Admittedly, this change doesn't make things _worse_, but it does give the appearance that this code is thread safe when it is not.

There's similar non-thread safe access to phaseMap on line 112 where an entry is removed.

Reply:
I am actually glad that you brought up this point. The code is actually thread safe, because of the distinct keys across threads. So, essentially we need to ensure that there is no corruption while the metadata is getting updated for the underlying map. Using `ConcurrentHashMap` implementation ensures that, and the same is reflected by tests as well.

```
// Make phase name unique for concurrent slices by including thread info
String uniqueElement = element + "_" + Thread.currentThread().getId();
```

Reply:
Ah I see, makes sense.